samplers/aws/xray/remote_sampler.go

```diff
@@ -0,0 +1,221 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package xray
+
+import (
+	"context"
+	crypto "crypto/rand"
+	"errors"
+	"fmt"
+	"log"
+	"sync"
+	"time"
+
+	"go.opentelemetry.io/otel/sdk/trace"
+)
+
+// RemoteSampler is an implementation of SamplingStrategy.
+type RemoteSampler struct {
+	// List of known centralized sampling rules
+	manifest *centralizedManifest
+
+	// proxy is used for getting quotas and sampling rules
+	xrayClient *xrayClient
+
+	// pollerStart, if true represents rule and target pollers are started
+	pollerStart bool
+
+	// Unique ID used by XRay service to identify this client
+	clientID string
+
+	// Provides system time
+	clock Clock
+
+	mu sync.RWMutex
+}
+
+// Compile time assertion that remoteSampler implements the Sampler interface.
+var _ trace.Sampler = (*RemoteSampler)(nil)
+
+// NewRemoteSampler returns a centralizedSampler which decides to sample a given request or not.
+func NewRemoteSampler() *RemoteSampler {
+	// Generate clientID
+	var r [12]byte
+
+	_, err := crypto.Read(r[:])
+	if err != nil {
+		log.Println("error reading cryptographically secure random number generator")
+		return nil
+	}
+
+	id := fmt.Sprintf("%02x", r)
+
+	clock := &DefaultClock{}
+
+	m := &centralizedManifest{
+		rules: []*centralizedRule{},
+		index: map[string]*centralizedRule{},
+		clock: clock,
+	}
+
+	return &RemoteSampler{
+		pollerStart: false,
+		clock:       clock,
+		manifest:    m,
+		clientID:    id,
+		// ToDo: take proxyEndpoint and pollingInterval from users
+		xrayClient: newClient("127.0.0.1:2000"),
+	}
+}
+
+func (rs *RemoteSampler) ShouldSample(parameters trace.SamplingParameters) trace.SamplingResult {
+	// ToDo: add business logic for remote sampling
+	rs.mu.Lock()
+	if !rs.pollerStart {
+		rs.start()
+	}
+	rs.mu.Unlock()
+
+	return trace.SamplingResult{}
+}
+
+func (rs *RemoteSampler) Description() string {
+	return "remote sampling with AWS X-Ray"
+}
+
+func (rs *RemoteSampler) start() {
+	if !rs.pollerStart {
+		rs.startRulePoller()
+	}
+
+	rs.pollerStart = true
+}
+
+func (rs *RemoteSampler) startRulePoller() {
+	// Initial refresh
+	go func() {
+		if err := rs.refreshManifest(); err != nil {
+			log.Printf("Error occurred while refreshing sampling rules. %v\n", err)
+		} else {
+			log.Println("Successfully fetched sampling rules")
+		}
+	}()
+
+	// Periodic manifest refresh
+	go func() {
+		// Period = 300s, Jitter = 5s
+		t := newTicker(300*time.Second, 5*time.Second)
+
+		for range t.C() {
+			if err := rs.refreshManifest(); err != nil {
```

Comment:
```suggestion
		for {
			if err := rs.refreshManifest(); err != nil {
			...
			}
			select {
			case _, more := <-t.C:
			  if !more {
			    break
			  }
			  continue
			}
		}
```

The two goroutines started here could be collapsed into one by rearranging the block on the channel read.

Since this eventually makes a network request there should probably also be a `context.Context` wired through this and the `select` should have a `case <-ctx.Done()`.